### PR TITLE
arm: core: Initial Stack Guarding implementation

### PR DESCRIPTION
**Pull request by fvincenzo**
This pull request is a follow up of https://gerrit.zephyrproject.org/r/#/c/12499/

### CODE REVIEW

**Review by dbkinder — COMMENTED** [review]
+1 docs look OK

File: arch/arm/soc/nxp_kinetis/k6x/Kconfig.soc
Comment (MaureenHelm) [inline arch/arm/soc/nxp_kinetis/k6x/Kconfig.soc]: This should be one level higher, in arch/arm/soc/nxp_kinetis/Kconfig, because the sysmpu exists on other parts outside the k6x series.

File: arch/arm/soc/nxp_kinetis/k6x/nxp_mpu_regions.c
Comment (MaureenHelm) [inline arch/arm/soc/nxp_kinetis/k6x/nxp_mpu_regions.c]: Per my gerrit review, I'm not sure this address is correct. I think it should be 0x1BFF_FFFF?

**Review by MaureenHelm — DISMISSED** [review]

File: arch/arm/core/cortex_m/mpu/nxp_mpu.c
Comment (MaureenHelm) [inline arch/arm/core/cortex_m/mpu/nxp_mpu.c]: The SCGC registers are SoC-specific, so this isn't going to work across SoCs. Take a look at ext/hal/nxp/mcux/drivers/fsl_sysmpu.c in the SYSMPU_Init() function for a more portable example to enable the clock. You should be able to do this:
```
CLOCK_EnableClock(kCLOCK_Sysmpu0);
```

File: boards/arm/frdm_k64f/frdm_k64f_defconfig
Comment (MaureenHelm) [inline boards/arm/frdm_k64f/frdm_k64f_defconfig]: Do we really want to enable this by default? By disabling execution from SRAM, we'll break anything that uses the kinetis flash driver. For example, mcuboot.

File: arch/arm/core/cortex_m/mpu/arm_mpu.c
Comment (dbkinder) [inline arch/arm/core/cortex_m/mpu/arm_mpu.c]: I'd recommend not using shortened urls, but the full url is longer than allowed by our coding standards:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0553a/Cihegaib.html
sigh.  I guess we're stuck with the shortened url :(

File: arch/arm/core/cortex_m/mpu/nxp_mpu.c
Comment (fvincenzo) [inline arch/arm/core/cortex_m/mpu/nxp_mpu.c]: Ok, I will do. Thanks

**Review by dbkinder — DISMISSED** [review]
+1 documentation

File: arch/arm/core/cortex_m/mpu/arm_mpu.c
Comment (pfalcon) [inline arch/arm/core/cortex_m/mpu/arm_mpu.c]: Let's update coding standards away from 1970ies VT100 terminal with 80 chars per line...

File: boards/arm/frdm_k64f/frdm_k64f_defconfig
Comment (fvincenzo) [inline boards/arm/frdm_k64f/frdm_k64f_defconfig]: I think that an application that has specific needs has to override this setting via its own config file. The default should be protecting against basic issues.

File: arch/arm/core/cortex_m/mpu/arm_mpu.c
Comment (fvincenzo) [inline arch/arm/core/cortex_m/mpu/arm_mpu.c]: Yes I think so. I used google at least it is a reliable service.

File: boards/arm/frdm_k64f/frdm_k64f_defconfig
Comment (mbolivar) [inline boards/arm/frdm_k64f/frdm_k64f_defconfig]: I think we need to discuss how to approach applications that want to read and write flash in general. Right now, Carbon is already broken with mcuboot for MPU-related reasons (in addition to https://github.com/zephyrproject-rtos/zephyr/pull/122).

Out of tree, we are maintaining patches like this to keep our own builds working (aside from the applications patches):

https://github.com/Linaro/zephyr/commit/14af77b21ce54e10a0c62139c988271f48776b23
https://github.com/Linaro/zephyr/commit/0b2a39cedf5ab1ed4d90efa361b14de89f347e06

I think the MPU support patches that are coming in are awesome -- they've already helped us catch bugs, some of which are in upstream Zephyr. But it would be good to discuss in particular what the plans are for applications which need to read and write flash and access "other" areas of memory like the UID bytes or system memory on STM32.

Comment (fvincenzo) [inline boards/arm/frdm_k64f/frdm_k64f_defconfig]: First of all thanks for the info, I was not aware of it. This is definitely something that we need to target in the near future and I am going to send to you a separate email so we can formalize the requirements. 

Said that the patch you are referring to with your changes https://github.com/zephyrproject-rtos/zephyr/commit/5951e45580c198c11e8fac59544195123a46e3ad has been in RFC for nearly 2 months and I was hoping to have captured all the requirements in terms of exceptions but it seems something it is still missing :-)

With respect to the NXP MPU configuration that is part of this specific patch we are not targeting the scenarios that you are describing hence in my previous comment I was suggesting that at this stage without specific requirements the way to use mcuboot like applications is overriding CONFIG_HAS_SYSMPU in the application configuration setting it to n.

The support of the scenarios you described will be part of a future patch once the scope and the usecases are formalized.

**Comment by mbolivar** [comment]
> Said that the patch you are referring to with your changes zephyrproject-rtos/zephyr@5951e45 has been in RFC for nearly 2 months and I was hoping to have captured all the requirements in terms of exceptions but it seems something it is still missing.

Yeah, I think we should have paid more attention to your efforts when they were at the RFC stage; hopefully it is better late than never.

> without specific requirements the way to use mcuboot like applications is overriding CONFIG_HAS_SYSMPU in the application configuration setting it to n.

I hope we can work together to try to capture some requirements here that will allow the MPU to be turned on even for mcuboot.

Let me in particular point out this patch that I had pasted earlier:

https://github.com/Linaro/zephyr/commit/14af77b21ce54e10a0c62139c988271f48776b23

We use this both in our mcuboot builds and for an application (https://github.com/linaro/zephyr-fota-hawkbit) which writes flash to download firmware over the air updates for mcuboot to use.

In current Zephyr, of course, this is the same as giving all code RW access to flash, since all code is privileged. But the intent is to try to allow privileged components of applications to read/write flash if they want to, without affecting the defaults that flash is read-only. This allows you to leave the MPU on, catching other bugs.

With the MPU on, we've so far seen:

- network stack issues in Zephyr (caught in the application)
- bad address calculations in mcuboot
- even bugs in the MPU configuration itself (caught in the application, addressed as https://github.com/Linaro/zephyr/commit/0b2a39cedf5ab1ed4d90efa361b14de89f347e06)

None of these would have been found if we had to turn the MPU off just because we wanted to write flash.

**Comment by fvincenzo** [comment]
Ok, agreed, I am going to add the requested features to both NXP and ARM MPUs based on the requirements formalized via e-mail.

Once I pushed the the changed, I would like to ask to you to test them in your setup and let me know if the re is still something outstanding. :smile:

File: samples/mpu_test/src/main.c
Comment (erwango) [inline samples/mpu_test/src/main.c]: Is that 16K or 128k?

Comment (fvincenzo) [inline samples/mpu_test/src/main.c]: Should be 16 here. Thanks for catching this.

**Comment by mbolivar** [comment]
Hi @fvincenzo; thanks for the update. I've tested this series on an mcuboot build for FRDM-K64F with CONFIG_MPU_ALLOW_FLASH_WRITE=y and verified that the bootloader is able to write to flash without causing an MPU violation.

**Comment by fvincenzo** [comment]
@mbolivar Happy to hear that! I hope this feature helps you guys to find all the bugs and make Zephyr solid rock! 😄

**Review by mike-scott — COMMENTED** [review]
Hi @fvincenzo,
Do you think it's possible to keep my authorship on the "Add Allow write on Flash" patch?  I know you changed the title and description a bit (and moved the Kconfig portion out into another patch), but essentially it's the same as: https://github.com/Linaro/zephyr/commit/14af77b21ce54e10a0c62139c988271f48776b23

**Comment by fvincenzo** [comment]
@mike-scott The reason why I have my authorship in the patch it is because I did not use any of your work directly (means I did not apply your patch on my tree). I implemented the patch based on the requirements coming from the e-mail thread.
Since you did some pre-existing work I think it is correct to have your sign-off-by in there which I will add tomorrow once in office. Is that ok for you?

**Comment by fvincenzo** [comment]
@mike-scott As promised I added your signed-off-by.

**Comment by agross-linaro** [comment]
@mike-scott @fvincenzo 

Looking at this code, there isn't any different way to add security options for this single region.  There is nothing novel about the implementation.  As such, a signed-off-by should be fine.

**Comment by mike-scott** [comment]
@fvincenzo thanks!
@agross-linaro : agreed :)

File: arch/arm/core/cortex_m/mpu/Kconfig
Comment (galak) [inline arch/arm/core/cortex_m/mpu/Kconfig]: we should separate this out into a different patch, since I don't see why we need a config MPU_ALLOW_FLASH_WRITE, can't we just do this if CONFIG_FLASH is enabled in the code?  Since that would imply you have the ability to write to the flash

File: include/arch/arm/cortex_m/mpu/nxp_mpu.h
Comment (galak) [inline include/arch/arm/cortex_m/mpu/nxp_mpu.h]: why not #if defined(CONFIG_FLASH)

**Review by galak — CHANGES_REQUESTED** [review]
I think we can just use CONFIG_FLASH and remove CONFIG_MPU_ALLOW_FLASH_WRITE

File: arch/arm/core/cortex_m/mpu/Kconfig
Comment (mike-scott) [inline arch/arm/core/cortex_m/mpu/Kconfig]: CONFIG_FLASH doesn't imply that you want to write to flash.  It needs to be an additional option.

Comment (galak) [inline arch/arm/core/cortex_m/mpu/Kconfig]: Why would you enable FLASH drivers unless you wanted to write to flash?

File: include/arch/arm/cortex_m/mpu/nxp_mpu.h
Comment (fvincenzo) [inline include/arch/arm/cortex_m/mpu/nxp_mpu.h]: Agreed we can change this.

File: arch/arm/core/cortex_m/mpu/Kconfig
Comment (mike-scott) [inline arch/arm/core/cortex_m/mpu/Kconfig]: Hrm.. now that I look, there isn't in a use-case in Zephyr for enabling CONFIG_FLASH that isn't write/erase centric.   Prior to adding a CONFIG_* option to adjust MPU settings, I was thinking of a dynamic MPU hook which could be triggered via flash_write_protection_set() calls.  But, I'm not sure if we can do that in a generic way across MPU implementations.  But the side effect of that method would be no additional CONFIG options.

**Comment by mbolivar** [comment]
One reason is to read from a flash device that isn't memory mapped

**Comment by galak** [comment]
Ok, than you can use the flash driver cfg itself, CONFIG_SOC_FLASH_MCUX

**Comment by fvincenzo** [comment]
In my opinion the configuration option should not be specific to the SOC or to the Flash Driver and the reason is because the application should be able to configure the behavior in the same way independently from the MCU is running on (I do not think we want to have a prj.conf for each MCU we are supporting in mcuboot). This is the reason why at the beginning I defined  CONFIG_MPU_ALLOW_FLASH_WRITE and agreed with changing it in CONFIG_FLASH (which should enable automatically the specific flash).

If CONFIG_FLASH does not cover all the scenarios, I can revert back to the original config option. Let me know what do you think.

File: samples/mpu_test/prj_allow_flash_write_mcux.conf
Comment (MaureenHelm) [inline samples/mpu_test/prj_allow_flash_write_mcux.conf]: You don't need to set CONFIG_SOC_FLASH_MCUX=y explicitly. The Kinetis SoC defconfigs enable it if CONFIG_FLASH=y

File: samples/mpu_test/README.rst
Comment (MaureenHelm) [inline samples/mpu_test/README.rst]: I ran this sample on frdm_k64f and noticed the console output doesn't match what is documented in the README.rst.  Looks like perhaps the shell got added after the doc was written? Should probably fix in a follow up patch

File: samples/mpu_test/src/main.c
Comment (MaureenHelm) [inline samples/mpu_test/src/main.c]: Better to use sizeof(value)?

**Comment by mbolivar** [comment]
> the application should be able to configure the behavior in the same way independently from the MCU is running on (I do not think we want to have a prj.conf for each MCU we are supporting in mcuboot)

From an mcuboot contributor perspective, I think this is spot on.

Note also that all Zephyr applications which are chain-loaded by mcuboot have to use their flash drivers to update some state shared with mcuboot, to e.g. signal that they booted correctly and don't need to be rolled back to an old version.

So if you make this SOC flash driver-specific, each of those applications will have to either disable the MPU or maintain a config file for every board they support, which doesn't scale. With the patch as-is, however, you can let the applications express their intent "I want to use the MPU, but I need to write my flash" in just one file.

**Comment by fvincenzo** [comment]
@mbolivar From what you are saying seems that your preference is CONFIG_MPU_ALLOW_FLASH_WRITE. Am I correct?

**Review by MaureenHelm — COMMENTED** [review]
> @mbolivar From what you are saying seems that your preference is CONFIG_MPU_ALLOW_FLASH_WRITE. Am I correct?

I think so, and I think I agree

File: samples/mpu_stack_guard_test/testcase.ini
Comment (andrewboie) [inline samples/mpu_stack_guard_test/testcase.ini]: why is this a build_only test?
There are ways of testing whether the system correctly caught an exception.
See tests/kernel/fatal for an example.
I'd like to see this moved to tests/ and verify correctness at runtime.

**Review by tyrantlizard77 — DISMISSED** [review]
Is it possible to introduce an MPU configuration for qemu_cortex_m3 so that we can test some of these MPU features with sanitycheck?

**Comment by galak** [comment]
I don't see why you end up with per MCU proj.conf files.  For a given SoC you will end up in the .config have both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled.  All that I'm saying is in nxp_mpu.h we'd use the CONFIG_SOC_FLASH_MCUX.

**Comment by mike-scott** [comment]
@galak I think as @mbolivar pointed out reading from external (or non-memory mapped) flash would require enabling CONFIG_FLASH.  And that use-case doesn't equate to wanting to write to flash.

So, I agree with the others that a separate CONFIG to allow MPU access for flash writes is the best option for the moment.

**Comment by mbolivar** [comment]
> I don't see why you end up with per MCU proj.conf files. For a given SoC you will end up in the .config have both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled. All that I'm saying is in nxp_mpu.h we'd use the CONFIG_SOC_FLASH_MCUX.

Hm, I must have misunderstood your proposal. Let me explain myself better.

To be clear, I do think CONFIG_MPU_ALLOW_FLASH_WRITE is worth having.

Here are the cases I can see:

1. User does not want to use the MPU, and may or may not want the flash driver
2. User wants to use the MPU, but does not want to write flash.
2.1. Sub-case where user wants the SoC flash driver, but to read flash only (I know that it's memory-mapped in this case and you could just read with load instructions, but it's better to use the driver.)
2.2. Sub-case where user does not want the SoC flash driver at all.
3. User wants MPU, and does want to write flash. In this case, the user wants the SoC flash driver, for reading and writing.

Here's how they could be supported in the presence of MPU_ALLOW_FLASH_WRITE.

1. MPU config = n, MPU_ALLOW_FLASH_WRITE isn't meaningful, SoC flash driver config = don't care
2.1. MPU config = y, MPU_ALLOW_FLASH_WRITE = n, SoC flash driver config = y
2.2 MPU config = y, MPU_ALLOW_FLASH_WRITE = n, SoC flash driver config = n
3. MPU config = y, MPU_ALLOW_FLASH_WRITE = y, SoC flash driver config = y

Can you explain how you'd support these without MPU_ALLOW_FLASH_WRITE?

**Comment by galak** [comment]
I don't see why anyone would realistically use the flash drive to read.  Why is it better to use the driver, its just doing load/store for flashes that are memory mapped.  So what's better about this?

Just enumerating all cases of MPU enabled/disabled + MPU_ALLOW_FLASH_WRITE enabled/disable doesn't really justify the need.  3 or the 4 cases can be handled as I suggested.  The only case is for some reason of having the MPU enabled but not wanting to allow flash writing but enabling the driver, this seems broken to me as we now cause MPU errors for what is a reasonable API to call.  If you enable the FLASH I think its a pretty reasonable expectation that you should be able to write to it.

**Comment by galak** [comment]
@mike-scott I agree that CONFIG_FLASH isn't sufficient, thus suggesting using the driver CONFIG option in the MPU code instead.

Additionally, even if we introduce CONFIG_MPU_ALLOW_FLASH_WRITE, I don't think it should be user selectable.

**Comment by galak** [comment]
@andrewboie we've looked at MPU support for QEMU, problem is MMU support on QEMU which is the obvious thing to base this on assumes a 4k page size (or at least that the smallest is my understanding).  So trying to support an MPU would require huge effort in QEMU.

**Comment by andrewboie** [comment]
@galak ah, makes sense. So qemu-system-arm only emulates the MMU and not any of the various MPUs out there. Bummer!

**Comment by mbolivar** [comment]
> I don't see why anyone would realistically use the flash drive to read. Why is it better to use the driver, its just doing load/store for flashes that are memory mapped. So what's better about this?

If that is true, why is there a flash_read() function at all?

Having a flash driver lets applications read from flash without having to duplicate the code for where the base address is, etc. The results are more self-documenting, for one, and are also useful for applications which may in some configurations want to only read flash and in others to read and write.

If you want an example of such an application: mcuboot itself. There was a lot of feedback at the last Connect that expressed discomfort at how much work it was doing (A/B swapping being safe to do in the face of random power failures was the main complaint). Since then, we've been looking seriously at an mcuboot configuration that only reads flash, and picks a 'most recent' or 'best' application to boot.

If we can have separate CONF_FILEs that let us switch which boot style (A/B swap with read/write vs. boot only) and simultaneously change the MPU configuration (MPU+write flash vs. MPU+never write flash), then the same code can be reused to read image headers, run checksums, etc.

If that's not possible, then the "read-only" mcuboot has to basically include its own copy of a generic flash driver. If any hardware ever comes along that has quirky flash reads, then it'll have to copy those quirks as well.

So I think there's already a good reason to have this, and for it to be user selectable.

File: samples/mpu_test/prj_allow_flash_write_mcux.conf
Comment (fvincenzo) [inline samples/mpu_test/prj_allow_flash_write_mcux.conf]: Thanks I will remove it.

File: samples/mpu_test/README.rst
Comment (fvincenzo) [inline samples/mpu_test/README.rst]: Yes, right this sample changed a bit over time and I forgot to update the output.

File: samples/mpu_test/src/main.c
Comment (fvincenzo) [inline samples/mpu_test/src/main.c]: Ok, I can change it.

File: samples/mpu_stack_guard_test/testcase.ini
Comment (fvincenzo) [inline samples/mpu_stack_guard_test/testcase.ini]: Thanks for pointing this out. I will adapt my code accordingly.

**Comment by fvincenzo** [comment]
> I don't see why you end up with per MCU proj.conf files. For a given SoC you will end up in the .config have both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled.

At this stage of the Zephyr development this seems to be true only for the nordic_nrf5, the kinetis family and the arduino 101. Is that correct?

**Comment by mbolivar** [comment]
@galak what if CONFIG_MPU_ALLOW_FLASH_WRITE is default y if CONFIG_FLASH?

That should keep the default sane for users that load the flash driver, but still allow users to test that their applications which read flash aren't erasing or writing it if they don't want them to.

**Comment by galak** [comment]
So my feeling is to make progress we remove CONFIG_MPU_ALLOW_FLASH_WRITE for right now.  Than we can introduce a new PR which uses CONFIG_MPU_ALLOW_FLASH_WRITE once this PR gets merged.

**Comment by MaureenHelm** [comment]
> both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled. All that I'm saying is in nxp_mpu.h we'd use the CONFIG_SOC_FLASH_MCUX.

While I'm not opposed to CONFIG_MPU_ALLOW_FLASH_WRITE, this works too.

The k64 related changes look good.

**Comment by mbolivar** [comment]
No issues from me; we can carry that patch out of tree.

**Review by galak — APPROVED** [review]

**Pull request merged**
